### DISCUSSION TITLE
Invisible viewports

### DISCUSSION THREAD

**Discussion by gswifort** [CEXT-Dan/PyRx#98] [Q&A]
This code returns me more viewports than I actually have in the drawing. I have one sheet with one viewport and I get e.g. 4
Can I somehow check if the viewport is in the paper space? iterating over the layout also returns more viewports.

```py
vps = [Db.Viewport(id) for id in db.objectIds(Db.Viewport.desc())]
```

**Comment by gswifort** [comment]
I have established that:
- Autocad and Zwcad always have one additional viewport for each sheet (it is returned as the last by `.objectIds()` so you can always skip the last element)
- Autocad removes viewports when deleting a sheet, while Zwcad keeps the viewports of removed sheets (only when the file is restarted are the viewports of non-existent sheets removed).

**Reply by CEXT-Dan** [reply]
db.objectIds is kind of a sledgehammer as it iterates the handles. There’s a chance you can pick up orphaned objects, although they should be marked as effectively erased. It could be the ZwCad has some cobwebs, but not necessarily bugs 
also, you will also pickup the modelspace viewport 

there’s  Db.curDb().getViewportArray(), but I noticed I have a bug(fixed) in that it ignored the bool flag
here you will have to get the parent(s) to find the layouts your on

if you work though the layout dictionary, you can maintain the layout – viewport relationship
 
```Python
def PyRxCmd_doit():
    try:
        db = Db.curDb()
        ld = Db.Dictionary(db.layoutDictionaryId())
        for key, value in ld.asDict().items():
            layout = Db.Layout(value)
            if layout.getBlockTableRecordId() == db.modelSpaceId():
                continue
            
            #The first ID in the list will be the paperspace viewport
            print(key, len(layout.getViewportArray()))
    except Exception as err:
        traceback.print_exc(err)
```
